src/workerd/io/worker.c++

```diff
@@ -2595,7 +2599,15 @@ private:
     // Message pumping promise that should be evaluated on the InspectorService
     // thread.
     kj::Promise<void> messagePump() {
-      return receiveLoop().exclusiveJoin(dispatchLoop()).exclusiveJoin(transmitLoop());
+      // Although inspector I/O must happen on the InspectorService thread (to make sure breakpoints
+      // don't block inspector I/O), inspector messages must be actually dispatched on the Isolate
+      // thread. So, we run the dispatch loop on the Isolate thread.
+      auto dispatchLoopPromise = isolateThreadExecutor->executeAsync([this]() {
```

Comment:
nit: I would move the executeAsync into the dispatchLoop method, it's a bit clearer for future readers imho. 
otherwise lgtm

Reply:
I felt the same intuition, but wasn't confident the result was actually more readable, given the infamous footgunnery of lambda coroutines. The main choices that I see are:

1. Rename `dispatchLoop()` to `dispatchLoopImpl()` and make a new `dispatchLoop()` whose body is a call to `dispatchLoopImpl()` inside a lambda passed to `executeAsync()`.
2. Or, move `dispatchLoop()`'s body into a constexpr lambda which accepts `incomingQueueNotifier` and `channel` as parameters. (Alternatively, it could accept a `WebSocketIoHandler& self parameter`, and `self.` qualify the member accesses.)
3. Or, move `dispatchLoop()`'s body into a non-constexpr lambda which captures `this`, and wrap the lambda in `kj::coCapture()`.
4. Or, convert `dispatchLoop()` to a recursive `.then()` function, which I think would also require a separate `dispatchLoopImpl()`, since we only want to call `executeAsync()` a single time.
5. Or, the PR as written.

Reply:
Not sure what you mean, seeing as how `this` (which is just a ptr) should outlive the promise returned by executeAsync there should be no difference between the current implementation and 
```c++
// Must be called on the Isolate thread.
kj::Promise<void> dispatchLoop() {
  return isolateThreadExecutor->executeAsync([this]() {
    for (;;) {
      co_await incomingQueueNotifier->awaitNotification();
      KJ_IF_SOME(c, channel) {
        co_await c.dispatchProtocolMessages(this->incomingQueue);
      }
    }
  });
}
```

Reply:
Ah, you are right that that alternative actually is safe, because `executeAsync()` saves the lambda for the duration of the promise it returns, which is a guarantee I forgot it provides: https://github.com/capnproto/capnproto/blob/6cb55ac7d523c24047e6bbef0679027561f3c0d4/c%2B%2B/src/kj/async.h#L1040-L1041

To make sure we're on the same page, though: my understanding is that `this` being a pointer does not make this inherently safe. That is, my understanding is that it's the lifetime of the captured pointers/references (i.e., the lifetime of the lambda), not the lifetime of the pointed-to objects (e.g. `*this`), that matters.

My fear is that someone might see a `this`-capturing lambda coroutine and assume it is a generally safe pattern to use, and pass such a thing to, say, `Promise<T>::then()`, which does not preserve the lambda's lifetime. Because of this, I'd much prefer a blanket ban against non-constexpr, non-`coCapture()`ed coroutine lambdas, even if some usages like this one can be assumed to be safe.